docs/decisions/0011-kernel-service-registration.md

```diff
@@ -0,0 +1,180 @@
+
+Pros:
+
+- No dependency on specific DI container library.
+- Lightweight implementation.
+- Possibility to register only those services that can be used by Plugins (isolation from host application).
+- Possibility to register same interface multiple times by **name**.
+
+Cons:
+
+- Implementation and maintenance for custom DI container, instead of using already existing libraries.
+- To import Plugin, it still needs to be initialized manually to inject specific service.
+
+### Solution #2.2
+
+This solution is an improvement for last disadvantage of Solution #2.1 to handle case, when Plugin instance should be initialized manually. This will require to add new way how to import Plugin into Kernel - not with object **instance**, but with object **type**. In this case, Kernel will be responsible for `TextMemoryPlugin` initialization and injection of all required dependencies from custom service collection.
+
+```csharp
+// Instead of this
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
+
+// Use this
+kernel.ImportFunctions<TextMemoryPlugin>();
+```
+
+### Solution #3
+
+Instead of custom service collection and service provider in Kernel, use already existing DI library - `Microsoft.Extensions.DependencyInjection`.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+
+serviceCollection.AddTransient<IMemoryStore, VolatileMemoryStore>();
+serviceCollection.AddTransient<ITextEmbeddingGeneration>(
+    (serviceProvider) => new OpenAITextEmbeddingGeneration(modelId, apiKey));
+
+serviceCollection.AddTransient<ISemanticTextMemory, SemanticTextMemory>();
+
+var services = serviceCollection.BuildServiceProvider();
+
+var kernel = Kernel.Builder
+    .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+    .WithOpenAITextEmbeddingGenerationService(modelId, apiKey)
+    .WithServices(services) // Pass all registered services from host application to Kernel
+    .Build();
+
+// Plugin Import - option #1
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
+
+// Plugin Import - option #2
+kernel.ImportFunctions<TextMemoryPlugin>();
+```
+
+Pros:
+
+- No implementation is required for dependency resolution - just use already existing .NET library.
+- The possibility to inject all registered services at once in already existing applications and use them as Plugin dependencies.
+
+Cons:
+
+- Additional dependency for Semantic Kernel package - `Microsoft.Extensions.DependencyInjection`.
+- No possibility to include specific list of services (lack of isolation from host application).
+- Possibility of `Microsoft.Extensions.DependencyInjection` version mismatch and runtime errors (e.g. users have `Microsoft.Extensions.DependencyInjection` `--version 2.0`  while Semantic Kernel uses `--version 6.0`)
```

Comment:
This is only a problem if there are breaking changes between the versions, which we try really hard to avoid; otherwise, package resolution will handle picking an appropriate version everyone's happy with.

This is also the case for every dependency SK has, whether it be LoggingFactory or System.Text.Json or whatever.  DI isn't really special here.